### PR TITLE
CURATOR-705: Add additional test case to TestServiceCache

### PR DESCRIPTION
**Pull request by naude-r**
Highlight the potential issues as described in CURATOR-705.

### CODE REVIEW

File: curator-x-discovery/src/test/java/org/apache/curator/x/discovery/TestServiceCache.java [hunk 1]
```diff
@@ -114,6 +117,81 @@ public void stateChanged(CuratorFramework client, ConnectionState newState) {}
...
+                    .port(10065)
+                    .build();
+            ServiceInstance<String> instance3 = ServiceInstance.<String>builder()
+                    .payload("test")
+                    .name("test")
+                    .port(10066)
+                    .build();
+            discovery.registerService(instance2);
+            discovery.registerService(instance3);
+            assertEquals(3, cache.getInstances().size());
```
Comment (kezhuw) [inline curator-x-discovery/src/test/java/org/apache/curator/x/discovery/TestServiceCache.java:177]: Adding `Thread.sleep(500)` before assertions pass them.

I think it is about freshness, and the javadoc says:

>  NOTE: there is no guarantee of freshness.

Similar, `testInitialLoad` asserts after `latch.await`.

**Review by kezhuw — COMMENTED** [review]
I left comments inline related to concrete assertions.

Regarding to the `CuratorFramework::runSafe`, it is only used for notification purpose. No hard works are run on it. I changed the executor to `Executors.newSingleThreadExecutor()`, and all behave well.

File: curator-x-discovery/src/test/java/org/apache/curator/x/discovery/TestServiceCache.java (hunk 1 above)
Comment (kezhuw) [reply curator-x-discovery/src/test/java/org/apache/curator/x/discovery/TestServiceCache.java:177]: Swapping the order of assertion and `latch.await` is sufficient for this to pass. Then it is same as `testInitialLoad`.

Comment (naude-r) [reply curator-x-discovery/src/test/java/org/apache/curator/x/discovery/TestServiceCache.java:177]: @kezhuw thank you for taking the time. in a real application we cannot afford to wait and we do not know if any instances are available. failing faster would be better.

you are right about the freshness. guess we were spoiled with ` PathChildrenCache::start(BUILD_INITIAL_CACHE)` and the synchronicity thereof.

Comment (kezhuw) [reply curator-x-discovery/src/test/java/org/apache/curator/x/discovery/TestServiceCache.java:177]: `ServiceCache::start` is a blocking operation, immediate `getInstances` reflects initial cache after start. `ServiceCacheListener::cacheChanged` will be called every time there is a change after start.

**Comment by kezhuw** [comment]
Close with same reason as jira.

But I think we can improve the javadoc a bit to state:
* `ServiceCache::start` blocks until fully initialized with all existing services for `getInstances`.
* `ServiceCacheListener::cacheChanged` will be called every time there is a change after `ServiceCache::start`.

**Pull request closed without merging**